Add baseURI leak from sandboxed iframes (#169)
xsleaks/wiki: content/docs/attacks/baseuri.md

+++

> title = "baseURI"

Reviewer: ```suggestion
title = "Sandboxed iframes"
```

description = ""
date = "2024-08-20"
category = [
    "Attack",
]
abuse = [
    "iframes",
]

> defenses = [
>     "Browser Fix",
>     "Application Fix",
> ]

Reviewer: I'd make it empty 

```suggestion
```

menu = "main"
weight = 3
+++

> `document.baseURI` can be used in an opaque-origin sandboxed iframe to leak the full URL of an ancestor page.
> 
> If a URL has sensitive information, the origin, query, or fragment (hash) are most likely to contain this sensitive information.

Reviewer: I'd remove the intro for now, since we might extend the section with new leaks for sandboxed iframes, e.g. insufficient process isolation. 
```suggestion
```

> ## Sandboxed opaque-origin about:srcdoc iframe

Reviewer: ```suggestion
## Base URL
```

> An iframe loaded with `about:srcdoc` and sandboxed without `allow-same-origin` (i.e. has opaque origin) can read `document.baseURI` to leak the closest http(s):// origin document's URL.

Author: Thanks for the rewrite. It's a bit more nuanced in HTML spec + Chromium (need to check Firefox), since `about:blank` behaves differently than `about:srcdoc`. I'll work from your rewrite and expand this to explain behaviors separately in separate commit.

> This also works in nested frames, with the baseURI value set to the closest document's URL that has an http(s):// origin. For example, nesting multiple `about:srcdoc` within `https://example.com/path?query#hash` will still leak the full `example.com` URL.

Author: This isn't strictly true for `about:srcdoc`. It's more nuanced (by my count there's about 4 possible behaviors), so I'll expand on this in separate commit. I should have expanded on this from the first commit, so will do so in separate commit.

> ## Code Snippet

Reviewer: ```suggestion
### Code Snippet
```

Adapted from crbug 40867031[^crbug-40867031]:
1. Navigate to a URL, where the URL contains secrets. e.g. https://example.com/path?query#hash
2. Run the following JavaScript in DevTools:
```javascript
f = document.createElement("iframe");
f.sandbox = "allow-scripts";
f.srcdoc = "<script>document.write('origin: ' + origin + ', baseURI: ' + document.baseURI);</script>";
document.body.appendChild(f);
```

{{< hint info >}}
Technically this also works with `about:blank` (verified via DevTools), but only an extension might be able to script this, so it's not that useful.
{{< /hint >}}

> ## Verified Browser Versions

Reviewer: ```suggestion
```

As of August 20th, 2024:
* Chrome 127.0.6533.120 Stable + 129.0.6668.9 Canary
* Edge 127.0.2651.105 Stable

> * Firefox 128.0 Stable

Reviewer: We usually add a hintbox informing about what browsers are affected. See an example in https://xsleaks.dev/docs/attacks/cache-probing/#defense

Author: Will update in separate commit.

## Defense

> * HTML specification fix [^html-spec-9025]
> * Browser fixes: [^crbug-40867031] [^crbug-330744612] 
> * Application mitigation: Applications should avoid having sensitive information in URL if the page may include sandboxed `about:srcdoc` iframes with untrusted data.

Reviewer: ```suggestion
Applications should avoid using opaque iframes to render untrusted code. Browsers are researching ways of mitigating the issues with sandboxed iframes [^html-spec-9025] [^crbug-40867031] [^crbug-330744612].
```

Author: Opaque sandboxed iframes are relatively safe to use in many cases, such as iframes with content loaded from a separate domain (e.g. `googleusercontent.com`). Origins can still be leaked via `location.ancestorOrigins`, but full URLs can be protected.

Does the following work better, to scope mitigations based on what needs to be protected?
> To prevent leaking full URLs, applications should avoid using opaque iframes with `about:srcdoc` to render untrusted code.
To prevent leaking URL origins, applications should avoid using opaque iframes to render untrusted code.
Browsers are researching ways of mitigating the issues with sandboxed iframes [^html-spec-9025] [^crbug-40867031] [^crbug-330744612].

Reviewer: I generalized the statement to also cover risks from `data:`, sandboxed `blob:` and `srcdoc`. The word opaque isn't well defined anywhere, but opaque usually refers to documents not loaded from network.
